iothub/device/devdoc/API Design.md

```diff
@@ -14,72 +15,79 @@ public Task<Properties> GetPropertiesAsync(CancellationToken cancellationToken =
+/// <param name="propertyConvention">A convention handler that defines serializer to use for the properties.</param>
+/// <param name="componentName">The component name this property belongs to.</param>
 /// <param name="cancellationToken">A cancellation token to cancel the operation.</param>
-public Task UpdatePropertiesAsync(IDictionary<string, object> properties, string componentName = default, IConventionHandler conventionHandler = default, CancellationToken cancellationToken = default);
+public Task UpdatePropertiesAsync(IDictionary<string, object> properties, PropertyConvention propertyConvention, string componentName = default, CancellationToken cancellationToken = default);
 
 /// <summary>
-/// Sets the global listener for Writable properties
+/// Update a writable property.
 /// </summary>
-/// <param name="callback">The global call back to handle all writable property updates.</param>
-/// <param name="userContext">Generic parameter to be interpreted by the client code.</param>
-/// <param name="cancellationToken">A cancellation token.</param>
-public Task SubscribeToWritablePropertyEvent(Action<Properties, object> callback, object userContext, CancellationToken cancellationToken = default);
+/// <param name="propertyName">Property name.</param>
+/// <param name="writablePropertyResponse">The writable properyt response to push.</param>
+/// <param name="componentName">The component name this property belongs to.</param>
+/// <param name="cancellationToken">A cancellation token to cancel the operation.</param>
+public Task UpdateWritablePropertyAsync(string propertyName, WritablePropertyResponse writablePropertyResponse, string componentName = default, CancellationToken cancellationToken = default);
+
+/// <summary>
+/// Update properties.
+/// </summary>
+/// <param name="propertyPatch">Reported properties to push.</param>
+/// <param name="cancellationToken">A cancellation token to cancel the operation.</param>
+public Task UpdatePropertiesAsync(PropertyCollection propertyPatch, CancellationToken cancellationToken = default);
 
 /// <summary>
 /// Sets the global listener for Writable properties
 /// </summary>
 /// <param name="callback">The global call back to handle all writable property updates.</param>
-/// <param name="componentName">The component name this writable property belongs to.</param>
 /// <param name="userContext">Generic parameter to be interpreted by the client code.</param>
-/// <param name="cancellationToken">A cancellation token.</param>
-public Task SubscribeToWritablePropertyEvent(Action<Properties, object> callback, string componentName = default, object userContext = default, CancellationToken cancellationToken = default);
+/// <param name="cancellationToken">A cancellation token to cancel the operation.</param>
+public Task SubscribeToWritablePropertyEventAsync(Func<PropertyCollection, object, Task> callback, object userContext, CancellationToken cancellationToken = default)
```

Comment:
I was looking through SubscribeToWritablePropertyEventAsync() and the corresponding PropertyCollection.  It’s not clear to me how the application is going to determine which component a given property is associated with?

If we only had Subscribe where the app had to specify the component name up front and the SDK routed appropriately, we wouldn’t need componentName in this collection.  But since we just have global handler the componentName does need to be part of data deserialized out.  

So not sure if we need some extra handling or i'm missing something?

Reply:
The component name will be a part of the PropertyCollection returned, it will just not have a unique dedicated accessor. What I mean is, if you iterate through a PropertyCollection returned, they key will always be either a root-level property name or a component name. The application will need to use a combination of switch cases to route their own internal callbacks in their dispatcher.
PropertyCollection:
-> root-property1
-> root-property2
-> component1
    -> component-property1
    -> component-property2
-> root-property3
...

This is how I imagine it to work.
I haven't updated the sample to include a root-level property update yet, but I am planning on adding it in today. I'll share a link once I am done and we can take a look.

Reply:
OK - provided there's a mechanism that's good.  And I like that you don't have an accessor, so you can reuse this same object in `UpdatePropertiesAsync` and let it set the componentName in parameter list of function.   Letting the componentName slip into the properties object gets more complicated.

Reply:
yeah, we do something like this:

```csharp
PropertyCollection propertyPatch;
if (string.IsNullOrWhiteSpace(componentName))
{
    propertyPatch = new PropertyCollection(properties, propertyConvention);
}
else
{
    properties.Add(PropertyConvention.ComponentIdentifierKey, PropertyConvention.ComponentIdentifierValue);
    var componentProperties = new Dictionary<string, object>
    {
        { componentName, properties }
    };

    propertyPatch = new PropertyCollection(componentProperties, propertyConvention);
}
```

Reply:
the dispatcher:
https://github.com/Azure/azure-iot-sdk-csharp/blob/e4b3dae81924184093c59a3808d8a78e553cc1ae/iothub/device/samples/PnpDeviceSamples/TemperatureController/TemperatureControllerSample.cs#L290-L304

the caller:
https://github.com/Azure/azure-iot-sdk-csharp/blob/e4b3dae81924184093c59a3808d8a78e553cc1ae/iothub/device/samples/PnpDeviceSamples/TemperatureController/TemperatureControllerSample.cs#L125-L133